File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        def dfs(root, left_bound, right_bound) -> bool:
            if not root:
                return True
            
            if root.val > left_bound and root.val < right_bound:
                return dfs(root.left, left_bound, root.val) and dfs(root.right, root.val, right_bound)
            else:
                return False
            
        return dfs(root, float("-inf"), float("inf"))

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int ans = INT_MIN;

    int dfs(TreeNode* node) {
        if (!node) return 0;

        int leftmax = max(dfs(node->left), 0);
        int rightmax = max(dfs(node->right), 0);

        ans = max(ans, node->val + leftmax + rightmax);

        return node->val + max(leftmax, rightmax);
    }

public:
    int maxPathSum(TreeNode* root) {
        dfs(root);
        return ans;
    }
};